### ISSUE TITLE
API improvement ideas (usable behind @beta flag on npm)

### ISSUE THREAD

**Issue by supermacro** [supermacro/neverthrow#150]
Just writing this here publicly to see if others agree or disagree. 

I think it would be nice to have a utility function that has the following signature:

```typescript
Promise<Result<T, E>> => ResultAsync<T, E>
```

Since sometimes it's easier to write a function using `async/await` syntax. 

---

Additionally, I am also thinking it would be nice to have another utility function with the following signature:

```
Result<T, E> => ResultAsyncT, E>
```

I guess the above would be called `intoResultAsync` or something like that? 

Curious to hear what folks think.

**Comment by supermacro** [comment]
Here is a real life example where I would need to convert a `Promise<Result<T, E>>` into a `ResultAsync<T, E>` and where using `ResultAsync.fromPromise` is not practical:


```typescript
export const getAdmin = async (
  adminId: Admin['id']
): Promise<Result<Admin, RouteError>> => {
  try {
    const admin = await prisma.admin({
      id: adminId,
    })

    return admin ? ok(admin) : err(RouteError.NotFound)
  } catch (e) {
    return err(RouteError.Other)
  }
}

```

**Comment by paduc** [comment]
This is how I would handle your example:

```typescript
ResultAsync.fromPromise(prisma.admin({ id: adminId}), () => RouteError.Other)
           .andThen((admin) => admin ? ok(admin) : err(RouteError.NotFound));
```

I’m not sure about other use-cases where it would be useful to have ```Promise<Result<T,E>> => ResultAsync<T,E>```.

Same for ```Result => ResultAsync```, I’m curious see where it could be useful.

***

One need which I stumble upon fairly often is a way to combine multiple Result into a single one. 
Let’s say for instance that I want to retrieve two bits of information and both queries return a ResultAsync. I want to execute an action after both queries arrive and only if they succeed.
I would love to have something like ```ResultAsync.combine([resultA, resultB]).andThen(([valueA, valueB]) => ...)```

**Comment by supermacro** [comment]
Here's a before and after of a function that originally returned a `Promise<Result<T, E>>`:

### Before:

```typescript
export const registerSite = async (
  adminId: Admin['id'],
  hostname: string
): Promise<Result<Site, RouteError>> => {
  try {
    const sites = await prisma
      .updateAdmin({
        data: {
          sites: {
            create: {
              hostname,
              dns_tag: v4(),
            },
          },
        },
        where: {
          id: adminId,
        },
      })
      .sites()

    const newSite = sites[0]

    return ok(newSite)
  } catch (e) {
    if (e.code && e.code === '23505') {
      return err(RouteError.Conflict)
    }

    logger.warn('Query Error', 'createAdmin', e)
    return err(RouteError.Other)
  }
}
```

### After:

```typescript
export const registerSite = (
  adminId: Admin['id'],
  hostname: string
): ResultAsync<Site, RouteError> => {
  const addSiteForAdmin = () => 
    prisma.updateAdmin({
      data: {
        sites: {
          create: {
            hostname,
            dns_tag: v4(),
          },
        },
      },
      where: {
        id: adminId,
      },
    })
      .sites()
      .then(([ newSite ]) => newSite)

  return ResultAsync.fromPromise(
    addSiteForAdmin(),
    (e) => {
      if (isObj(e) && e.code && e.code === '23505') {
        return RouteError.Conflict
      }

      logger.warn('Query Error', 'createAdmin', e)
      return RouteError.Other
    }
  )
}
```

This doesn't seem bad at all to me. I'd be interested in seeming cases where this sort of refactor isn't practical and would benefit from a utility function as described in my original comment above.

**Comment by paduc** [comment]
@supermacro I just stumbled upon a use-case for ```Promise<Result> => ResultAsync```.

```
const queue = new Queue();
const ticket = queue.push(async () => await publish(something)) // publish returns a ResultAsync<T> so ticket returns a Promise<Result<T>> (because I await publish to transform it from ResultAsync to Result)

const resultAsync = ResultAsync.fromPromise(ticket).andThen(res => res)
// resultAsync is a ResultAsync<T>
```

So basically the solution is to use `fromPromise` + `.andThen(res => res)` to transform  `Promise<Result>` into `ResultAsync`.

I noticed something more. If `ResultAsync` implemented the `PromiseLike` interface, the `.then` could transform a `ResultAsync` into a `Result` automatically.
```
// without PromiseLike
const res = promise.then(makesResultAsync) // ==> res is Promise<ResultAsync>

// If I want a Promise<Result> I need to do
const res = promise.then(async () => await makesResultAsync()) // ==> res is Promise<Result>

// if ResultAsync implemented PromiseLike, I would have
const res = promise.then(makesResultAsync) // ==> res is Promise<Result>
```

Interesting don't you think ?
The compiler says I'm missing `catch` `[Symbol.toStringTag]` and `finally` to be a `PromiseLike`

**Comment by supermacro** [comment]
Released a `combine` utility function behind a `@beta` flag on npm!

Docs on combine: https://github.com/supermacro/neverthrow/releases/tag/v2.8.0-beta.0

Source code: https://github.com/supermacro/neverthrow/blob/master/src/utils.ts

Feedback always welcome!

**Comment by paduc** [comment]
Great work !

Reading [line 26](https://github.com/supermacro/neverthrow/blob/master/src/utils.ts#L26), it reminded me that I've had this type of problem as well and found a possible solution: have `ResultAsync` implement `PromiseLike<Result>`.

I opened a PR which proposes exactly this => #151 
This would enable us to do things like
```typescript
const allResult = await Promise.all([okAsync<string, Error>('1')])
```

**Comment by supermacro** [comment]
Awesome! I've gone ahead and merged #151.

I will go ahead and publish a new version behind the`beta` tag.

**Comment by supermacro** [comment]
@paduc release is live.

https://github.com/supermacro/neverthrow/releases/tag/v2.8.0-beta.1

**Comment by supermacro** [comment]
Here's an example right now that I think justifies having a `Promise<Result<T, E>> => AsyncResult<T, E>` function:


```typescript
const getSiteAndComments = async (siteId: string): Promise<Result<SiteAndComments, RouteError>> => {
  const [ siteResult, commentsResult ]= await Promise.all([
    getSingleSite(siteId),
    getSiteComments(siteId),
  ])
  
  if (commentsResult.isErr()) {
    return err(commentsResult.error)
  } else if (siteResult.isErr()) {
    return err(siteResult.error)
  }

  return ok({
    site: siteResult.value,
    comments: commentsResult.value,
  })
}
```

**Comment by paduc** [comment]
> Here's an example right now that I think justifies having a `Promise<Result<T, E>> => AsyncResult<T, E>` function:

I personally never use `Promise<Result>` in my code because if there's a `Result` inside a function, it means that I have control over what's happening inside and can make that function return a `ResultAsync`.
Also, once you have a `Promise`, the caller will have to `await` it and you push the untyped error problem up the call chain. 
My principle is to transform the `Promise`s into `ResultAsync` at the lowest level (third-party api / io) and use `ResultAsync` everywhere in my code.

For your example, in which I guess your point is to call both `getSingleSite` and `getSiteComments` simultaneously, I would have written
```typescript
const getSiteAndComments = async (siteId: string): ResultAsync<SiteAndComments, RouteError> => {
  return ResultAsync.fromPromise(Promise.all([
    getSingleSite(siteId),
    getSiteComments(siteId),
  ])).andThen(([siteResult, commentsResult]) => {
    if (commentsResult.isErr()) {
    return commentsResult
  }
  if (siteResult.isErr()) {
    return siteResult
  }

  return ok({
    site: siteResult.value,
    comments: commentsResult.value,
  })
}
``

**Comment by supermacro** [comment]
Closing for now. 

`combine` has been released in v3.0.0, however it currently only works on homogeneous lists (where all the `T` and `E` are the same). In a future release, `combine` could be refactored to allow for arbitrary `T` values (but keeping `E` the same to allow for short-circuit on first error just like `Promise.all` does).
